consensus/istanbul/backend/engine_test.go

```diff
@@ -506,3 +577,175 @@ func TestWriteCommittedSeals(t *testing.T) {
 		t.Errorf("error mismatch: have %v, want %v", err, errInvalidCommittedSeals)
 	}
 }
+
+func makeSnapshotTestConfigItems() []interface{} {
+	return []interface{}{
+		stakingUpdateInterval(1),
+		proposerUpdateInterval(1),
+		proposerPolicy(params.WeightedRandom),
+	}
+}
+
+func makeFakeStakingInfo(blockNumber uint64, keys []*ecdsa.PrivateKey, amounts []uint64) *reward.StakingInfo {
+	stakingInfo := &reward.StakingInfo{
+		BlockNum: blockNumber,
+	}
+	for idx, key := range keys {
+		addr := crypto.PubkeyToAddress(key.PublicKey)
+
+		pk, _ := crypto.GenerateKey()
+		rewardAddr := crypto.PubkeyToAddress(pk.PublicKey)
+
+		stakingInfo.CouncilNodeAddrs = append(stakingInfo.CouncilNodeAddrs, addr)
+		stakingInfo.CouncilStakingAmounts = append(stakingInfo.CouncilStakingAmounts, amounts[idx])
+		stakingInfo.CouncilRewardAddrs = append(stakingInfo.CouncilRewardAddrs, rewardAddr)
+	}
+	return stakingInfo
+}
+
+func checkInValidators(target common.Address, list []istanbul.Validator) bool {
+	for _, val := range list {
+		if target == val.Address() {
+			return true
+		}
+	}
+	return false
+}
+
+func isAllDemoted(minimumStaking uint64, stakingAmounts []uint64) bool {
+	for _, val := range stakingAmounts {
+		if val >= minimumStaking {
+			return false
+		}
+	}
+	return true
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts        []uint64 // test staking amounts of each validator
+		isIstanbulCompatible  bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidatorsNum int      // the number of expected validators
+		expectedDemotedNum    int      // the number of expected demoted validators
```

Comment:
Not necessarily change this PR, however, please consider my comment in other PR. In my opinion, it is better to list up the indexes of expected validators/demoted, instead of their total numbers. I think direct comparing the expected outputs to the actual outputs is the purpose of the unittest.

Reply:
Thanks for the comment. I will apply it on the next PR.